File: downman/staginglist.py
#
#       staginglist.py
#
#       Copyright 2010 Brett Mravec <brett.mravec@gmail.com>
#
#       This program is free software; you can redistribute it and/or modify
#       it under the terms of the GNU General Public License as published by
#       the Free Software Foundation; either version 2 of the License, or
#       (at your option) any later version.
#
#       This program is distributed in the hope that it will be useful,
#       but WITHOUT ANY WARRANTY; without even the implied warranty of
#       MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#       GNU General Public License for more details.
#
#       You should have received a copy of the GNU General Public License
#       along with this program; if not, write to the Free Software
#       Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
#       MA 02110-1301, USA.

import downloaders

class StagingList:
    downloads = []
    view = None

    def __init__ (self, downman):
        self.downman = downman

    def add_download (self, download):
        self.downloads.append (download)

        if self.view != None:
            self.view.add_download (download)

        download.set_state_cb (self.on_info_cb)

        download.start_get_info ()

    def remove_download (self, download):
        self.downloads.remove (download)

        if self.view != None:
            self.view.remove_download (download)

        download.set_state_cb (None)

    def update_download (self, download):
        if self.view != None:
            self.view.update_download (download)

    def update_all (self):
        for d in self.downloads:
            self.update_download (d)

    def on_info_cb (self, download):
        if download.links != None:
            self.remove_download (download)

            for link in download.links:
                d = downloaders.create_download (link, self.downman)
                if d != None:
                    self.add_download (d)
        else:
            self.update_download (download)

    def clear (self):
        while len (self.downloads):
            self.remove_download (self.downloads[0])

    def get_all (self):
        ds = []

        for d in self.downloads:
            ds.append (d)

        return ds

File: src/staginglist.cc
/*
 *      staginglist.cc
 *
 *      Copyright 2010 Brett Mravec <brett.mravec@gmail.com>
 *
 *      This program is free software; you can redistribute it and/or modify
 *      it under the terms of the GNU General Public License as published by
 *      the Free Software Foundation; either version 2 of the License, or
 *      (at your option) any later version.
 *
 *      This program is distributed in the hope that it will be useful,
 *      but WITHOUT ANY WARRANTY; without even the implied warranty of
 *      MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *      GNU General Public License for more details.
 *
 *      You should have received a copy of the GNU General Public License
 *      along with this program; if not, write to the Free Software
 *      Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
 *      MA 02110-1301, USA.
 */

#include "staginglist.h"

StagingList::StagingList ()
{

}

StagingList::~StagingList ()
{

}

void
StagingList::add_download (Download *d, Download *nextd)
{
    std::vector<Download*>::iterator iter;
    Download *diter = NULL;

    d->start_get_info ();

    if (nextd == NULL) {
        for (iter = downloads.begin (); iter < downloads.end (); iter++) {
            if (*iter == nextd) {
                downloads.insert (iter, d);
                diter = *iter;
                break;
            }
        }
    }

    if (diter != NULL) {
        downloads.push_back (d);
    }

    d->signal_state_changed ().connect (sigc::mem_fun (*this, &StagingList::download_update));

    add (d, diter);
}

void
StagingList::update_download (Download *d)
{
    update (d);
}

void
StagingList::remove_download (Download *d)
{
    std::vector<Download*>::iterator iter;

    for (iter = downloads.begin (); iter < downloads.end (); iter++) {
        if (*iter == d) {
            downloads.erase (iter);
            remove (d);
            break;
        }
    }
}

void
StagingList::reorder_download (Download *d, Download *nextd)
{
    std::vector<Download*>::iterator iter;
    Download *diter = NULL;

    for (iter = downloads.begin (); iter < downloads.end (); iter++) {
        if (*iter == d) {
            downloads.erase (iter);
            break;
        }
    }

    if (nextd == NULL) {
        for (iter = downloads.begin (); iter < downloads.end (); iter++) {
            if (*iter == nextd) {
                downloads.insert (iter, d);
                diter = *iter;
                break;
            }
        }
    }

    if (diter != NULL) {
        downloads.push_back (d);
    }

    reorder (d, nextd);
}

void
StagingList::download_update (Download *d, DownloadState state)
{
    update_download (d);
}
